doc/reproducibility.md

```diff
@@ -2,3 +2,66 @@
 
 Depending on your application, you may also need to set the [numpy random seed](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.seed.html) or the seeds for other sources of randomness (for a brief overview, see [here](https://pytorch.org/docs/stable/notes/randomness.html)), but Augraphy does not depend on these for random number generation.
+
+---
+
+# Per-process Randomness vs. In-process Randomness
+One common misconception about setting the random seed is that this will guarantee identical pipeline runs within the same Python process. On the contrary, setting the seed will guarantee identical pipeline runs between multiple invocations of the same Python script.
+
+For example, consider the following code:
+
+``` python
+from augraphy import *
+import random
+import cv2
+
+random.seed(0)
+
+img = cv2.imread("image.png")
+
+augmented1 = default_augraphy_pipeline().augment(img)["output"]
+augmented2 = default_augraphy_pipeline().augment(img)["output"]
+
+cv2.imwrite("augmented1.png", augmented1)
+cv2.imwrite("augmented2.png", augmented2)
+
+print(augmented1.shape == augmented2.shape and not np.bitwise_xor(augmented1,augmented2).any())
+```
+
+Running this script once will produce the two images `augmented1.png` and `augmented2.png` in the local directory, and print `False` because two successive pipeline executions within the same Python process are not equivalent.
+
+We can modify the script slightly to save the augmented images under different names:
+
+``` python
+...
+cv2.imwrite("augmented1_new.png", augmented1)
+cv2.imwrite("augmented2_new.png", augmented2)
+```
+
+Run the new script to generate `augmented1_new.png` and `augmented2_new.png`.
+
+We can now run the following script to check whether `augmented1.png` and `augmented1_new.png`, and `augmented2.png` and `augmented2_new.png` are identical:
+
+``` python
+import cv2
+import numpy as np
+
+c1 = cv2.imread("augmented1.png")
+c1_new = cv2.imread("augmented1_new.png")
+
+c2 = cv2.imread("augmented2.png")
+c2_new = cv2.imread("augmented2_new.png")
+
+print(c1.shape == c1_new.shape and not np.bitwise_xor(c1,c1_new).any())
+print(c2.shape == c2_new.shape and not np.bitwise_xor(c2,c2_new).any())
+```
+
+The print statements will report whether the shape of the images is the same, and whether there are any non-equal pixels between the two images.
+
+``` bash
+$ python3 testaugmenteds.py
+True
+True
+```
+
+Indeed, the images generated by pipeline executions in successive Python processes are identical, and fixing the random seed made this possible.
```

Comment:
With this, we just need to set a fixed random seed in the main file before calling augraphy pipeline to enable a consistent output?

Reply:
Great, let me test on it first, thanks.

Reply:
If we have additional dependencies, it would be better to chain them such that setting the standard Python random seed propagates to these other libs.  In other words, set the seed in these downstream dependencies by simply calling the random.random() to obtain a seed for use in those libs.  Doing this will ensure the RNG output is deterministic regardless of the source.

Reply:
Another complimentary idea Alex had was to include a seed parameter in the pipeline that has a default value, or it could be null if we don't want to set/reset it.  Then we are declaring the method for reproducible randomness within the library's interface so that it is more clear and discoverable.  When the seed is provided, then we will set it per the rules we have documented.

Reply:
Commit [2a6c9cd](https://github.com/sparkfish/augraphy/pull/205/commits/2a6c9cd366c4b56061787ecb5032d62cd2086d1b) updates the documentation and adds the seed parameter to the pipeline which in turn sets the `random`, `numpy`, and `cv2` seeds.

Reply:
> If we have additional dependencies, it would be better to chain them such that setting the standard Python random seed propagates to these other libs. In other words, set the seed in these downstream dependencies by simply calling the random.random() to obtain a seed for use in those libs. Doing this will ensure the RNG output is deterministic regardless of the source.

I think defining the random seed within pipeline is better because it will be quite troublesome if we need to find out each function and define random.random() as their seed (in the case of opencv or numpy). If we missed out some functions, we will need more effort to troubleshoot and search for those functions too.

Reply:
> I think defining the random seed within pipeline is better because it will be quite troublesome if we need to find out each function and define random.random() as their seed (in the case of opencv or numpy). If we missed out some functions, we will need more effort to troubleshoot and search for those functions too.

I agree: the seed should be optionally set in the pipeline (see most recent commit), but anyone using individual augmentations outside of the pipeline can still set the seeds manually if they need e.g. reproducible `Letterpress`.

Reply:
So i think we can merge this pull request?